### PR TITLE
GH-1470: Group restriction fix

### PR DESCRIPTION
**Pull request by mpapierski**
Ref #1470 

This PR is an improved version of #1485 (reverted in #1495) with an emphasis on upgrade testing. It introduces the concept of upgrade regression testing by persisting a global state from a recent release into the version control and reusing it while testing. This way EE can iterate faster to detect post-upgrade regressions.

The fix itself consists of changes to how groups are verified in `call_contract` API call and additionally makes improvements arguments types/length verification. For example, an entry point defined with `CLType::Option` parameters makes this argument not checked.

### CODE REVIEW

File: execution_engine/src/core/runtime/mod.rs [hunk 1]
```diff
@@ -1852,36 +1846,123 @@ where
...
-            .map(|a| a.cl_type())
-            .cloned()
-            .zip(args.to_values().into_iter().map(|v| v.cl_type()).cloned())
+        // This will skip arguments check for system contracts only. This code should be removed on
+        // next major version bump. Argument checks for system contract is still done during
+        // execution of a system contract.
+        if !self
+            .protocol_data()
+            .system_contracts()
+            .contains(&contract_hash)
```
Comment (mpapierski) [inline execution_engine/src/core/runtime/mod.rs:1885]: @henrytill this line skips entry point arguments check for system contracts only. We can't upgrade entry points at the next minor/patch version bump without changing the upgrade procedure, so I figured the safest thing to do is just to not check the arguments for these until the next major version bump. This PR is ready for the next major bump (which will install fixed entrypoints) and has regression tests in place to verify it.

File: execution_engine/src/core/runtime/mod.rs [hunk 2]
```diff
@@ -1852,36 +1846,118 @@ where
...
+            // Unused args are the ones that are passed, but are not present as required nor
+            // optionals.
+            let unused_args = &(&passed_args - &required_args) - &optional_args;
+
+            if !unused_args.is_empty() {
+                return Err(Error::UnusedArgumentsFound {
+                    required: required_args.len(),
+                    unused: unused_args.len(),
+                });
+            }
```
Comment (EdHastingsCasperLabs) [inline execution_engine/src/core/runtime/mod.rs:1925]: I don't think it is safe for us to error in this case; some 3rd parties may tail their passed args with extra marker or readback data that is not needed for the called logic but is used for some off chain post process triggered by inclusion of the deploy in a block.

**Review by EdHastingsCasperLabs — APPROVED** [review]
Generally LGTM, but see note on unused params

File: execution_engine/src/core/runtime/mod.rs (hunk 2 above)
Comment (mpapierski) [reply execution_engine/src/core/runtime/mod.rs]: Fair point. I removed this error condition, and added an extra test to ensure contracts can accept arbitrary arguments (not listed in entry point as extra markers; for example)

**Comment by henrytill** [comment]
I would like to rebase this branch against #1627 so that it is aligned with that work.

**Review by henrytill — CHANGES_REQUESTED** [review]
I would like to rebase this branch against #1627 so that it is aligned with that work.

**Review by henrytill — COMMENTED** [review]
Otherwise looks good.

**Pull request merged**